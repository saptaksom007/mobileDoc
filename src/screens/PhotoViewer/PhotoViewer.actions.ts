import { MediaResource } from 'api/types'

export const ActionTypes = {
  PHOTOVIEWER_REQUESTED: 'PHOTOVIEWER_REQUESTED',
  PHOTOVIEWER_SUCCEEDED: 'PHOTOVIEWER_SUCCEEDED',
  PHOTOVIEWER_FAILED: 'PHOTOVIEWER_FAILED',
  PHOTOVIEWER_PREVIEW_REQUESTED: 'PHOTOVIEWER_PREVIEW_REQUESTED',
  PHOTOVIEWER_PREVIEW_SUCCEEDED: 'PHOTOVIEWER_PREVIEW_SUCCEEDED',
  PHOTOVIEWER_PREVIEW_FAILED: 'PHOTOVIEWER_PREVIEW_FAILED',
  PHOTOVIEWER_POP: 'PHOTOVIEWER_POP',
  LOCAL_PHOTO_CACHED: 'LOCAL_PHOTO_CACHED',
  NAVIGATE_TO_PHOTO_VIEWER: 'NAVIGATE_TO_PHOTO_VIEWER',
}

export default class Actions {
  static photoViewerRequested(mediaResource: MediaResource) {
    return {
      type: ActionTypes.PHOTOVIEWER_REQUESTED,
      payload: { mediaResource },
    }
  }
  static photoViewerSucceeded(dataUrl: string, uuid: string) {
    return {
      type: ActionTypes.PHOTOVIEWER_SUCCEEDED,
      payload: { dataUrl, uuid },
    }
  }
  static photoViewerFailed(error: Error) {
    return {
      type: ActionTypes.PHOTOVIEWER_FAILED,
      payload: {
        error,
      },
    }
  }
  static photoViewerPreviewRequested(mediaResource: any) {
    return {
      type: ActionTypes.PHOTOVIEWER_PREVIEW_REQUESTED,
      payload: { mediaResource },
    }
  }
  static photoViewerPreviewSucceeded(
    dataUrl: string,
    uuid: string,
    timestamp?: number,
  ) {
    return {
      type: ActionTypes.PHOTOVIEWER_PREVIEW_SUCCEEDED,
      payload: { dataUrl, uuid, timestamp },
    }
  }
  static photoViewerPreviewFailed(error: Error) {
    return {
      type: ActionTypes.PHOTOVIEWER_PREVIEW_FAILED,
      payload: {
        error,
      },
    }
  }
  static photoViewerPop() {
    return {
      type: ActionTypes.PHOTOVIEWER_POP,
      payload: undefined,
    }
  }
  static localPhotoCached() {
    return {
      type: ActionTypes.LOCAL_PHOTO_CACHED,
    }
  }
  static navigateToPhotoViewer(mediaResource: MediaResource | string) {
    return {
      type: ActionTypes.NAVIGATE_TO_PHOTO_VIEWER,
      payload: mediaResource,
    }
  }
}
