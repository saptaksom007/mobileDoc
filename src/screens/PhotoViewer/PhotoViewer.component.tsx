import React, { Component } from 'react'

import { View, StatusBar } from 'react-native'
import TransformableImage from '@applications-developer/react-native-transformable-image'
import { BackButton } from 'components/BackButton/BackButton.component'
import { Layout } from 'constants/Layout'
import { connect } from 'react-redux'
import * as ScreenOrientation from 'expo-screen-orientation'
import { isIphoneX } from 'react-native-iphone-x-helper'

import Actions from './PhotoViewer.actions'
import styles from './PhotoViewer.styles'
import { Action } from 'redux'

import { Loader } from 'components/Loader/Loader.component'
import { Color } from 'constants/Color'

interface Props {
  dataUrl: string
  dispatch(action: Action): void
}

export class PhotoViewer extends Component<Props> {
  static navigationOptions = {
    headerShown: false,
  }

  async componentDidMount() {
    if (!isIphoneX()) {
      await ScreenOrientation.lockAsync(ScreenOrientation.OrientationLock.ALL)
    }
    StatusBar.setHidden(true)
  }

  async componentWillUnmount() {
    if (!isIphoneX()) {
      await ScreenOrientation.lockAsync(
        ScreenOrientation.OrientationLock.PORTRAIT_UP,
      )
    }
    StatusBar.setHidden(false)
  }

  render() {
    const { dataUrl, dispatch } = this.props
    return (
      <View style={styles.container}>
        {dataUrl ? (
          <TransformableImage
            source={{ uri: dataUrl }}
            style={styles.transformImageActive}
            enableScale
          />
        ) : (
          <Loader
            color={Color.white}
            containerProps={{ style: { backgroundColor: 'transparent' } }}
          />
        )}

        <BackButton
          onPress={() => dispatch(Actions.photoViewerPop())}
          visible
          color='#fff'
          containerStyle={{
            width: Layout.window.width / 12,
            height: Layout.window.height / 8,
          }}
        />
      </View>
    )
  }
}

function select(
  { photoviewer: { previews } }: any,
  {
    navigation: {
      state: {
        params: { uuid, url },
      },
    },
  }: any,
) {
  return { dataUrl: previews[uuid] ? previews[uuid].dataUrl : url }
}

export default connect(select)(PhotoViewer)
