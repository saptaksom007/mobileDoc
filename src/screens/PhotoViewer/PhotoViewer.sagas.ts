import * as Crypto from 'expo-crypto'
import * as FileSystem from 'expo-file-system'
import { put, takeEvery, call, takeLatest } from 'redux-saga/effects'
import { Actions as NavAction } from 'navigation/SagaNavigation'
import { ActionTypes as ChatActionTypes } from 'screens/Chat/Chat.actions'
import { getMedia } from 'utilities/commonViewerSaga'
import { messagingActionTypes } from 'common-docdok/lib/domain/messaging/actions'
import MediaCache from 'common-docdok/lib/utils/mediaCache'

import Actions, { ActionTypes } from './PhotoViewer.actions'
import { api } from 'common-docdok'
import { MediaResource } from 'api/types'

const Cache: any = new MediaCache()

function* fetchPhoto(action: any) {
  try {
    const preview = false
    yield getMedia(action, preview)
  } catch (err) {
    yield put(Actions.photoViewerFailed(err))
  }
}

function* gotoPhotoViewer(action: any) {
  try {
    const { dataUrl, mimeType, preview, uuid } = action.payload
    if (mimeType !== 'application/pdf' && !preview) {
      yield put(Actions.photoViewerSucceeded(dataUrl, uuid))
    }
  } catch (err) {
    yield put(Actions.photoViewerFailed(err))
  }
}

const getImageFileSystemKey = async (uuid: string) => {
  const hashed = await Crypto.digestStringAsync(
    Crypto.CryptoDigestAlgorithm.SHA256,
    uuid,
  )
  return `${FileSystem.cacheDirectory}${hashed}`
}

function* fetchPhotoPreview(action: {
  payload: { mediaResource: MediaResource }
  type: string
}) {
  try {
    const {
      mediaResource: { uuid },
    } = action.payload
    if (uuid) {
      const filesystemURI = yield call(getImageFileSystemKey, uuid)
      const metadata: FileSystem.FileInfo = yield call(
        FileSystem.getInfoAsync,
        filesystemURI,
      )
      if (metadata.exists) {
        yield put(Actions.photoViewerPreviewSucceeded(filesystemURI, uuid))
      } else {
        const { data: signedUrl } = yield call(api.getMediaRedirectUrl, uuid)
        const dowloadResult: FileSystem.DownloadResult = yield call(
          FileSystem.downloadAsync,
          signedUrl,
          filesystemURI,
        )
        yield put(Actions.photoViewerPreviewSucceeded(dowloadResult.uri, uuid))
      }
    } else {
      yield put(
        Actions.photoViewerPreviewFailed(new Error('no uuid for resource')),
      )
    }
  } catch (err) {
    yield put(Actions.photoViewerPreviewFailed(err))
  }
}

function* showPhotoPreview(action: any) {
  try {
    const { dataUrl, mimeType, uuid, preview } = action.payload
    if (mimeType !== 'application/pdf' && preview) {
      const timestamp = new Date().getTime()
      yield put(Actions.photoViewerPreviewSucceeded(dataUrl, uuid, timestamp))
    }
  } catch (error) {
    yield put(Actions.photoViewerPreviewFailed(error))
  }
}

function* cacheLocalPhoto(action: any) {
  try {
    const { mediaResourceId, dataUri, uri } = action.payload
    if (dataUri) {
      yield call(Cache.set, mediaResourceId, uri)
      yield put(Actions.localPhotoCached())
    }
  } catch (error) {
    yield put(Actions.photoViewerPreviewFailed(error))
  }
}

function* pop() {
  yield put(NavAction.pop())
}

function* navigateToPhotoViewer({ payload }: any) {
  if (typeof payload === 'string') {
    yield put(NavAction.push('photoviewer', { url: payload }))
  } else {
    const { uuid } = payload
    yield put(NavAction.push('photoviewer', { uuid }))
  }
}

export default [
  takeEvery(ActionTypes.PHOTOVIEWER_REQUESTED, fetchPhoto),
  takeEvery(ActionTypes.PHOTOVIEWER_PREVIEW_REQUESTED, fetchPhotoPreview),
  takeEvery(ActionTypes.PHOTOVIEWER_POP, pop),
  takeEvery(messagingActionTypes.GET_MEDIA_RESOURCE_SUCCEEDED, gotoPhotoViewer),
  takeEvery(
    messagingActionTypes.GET_MEDIA_RESOURCE_SUCCEEDED,
    showPhotoPreview,
  ),
  takeEvery(ChatActionTypes.ADD_PHOTO_SUCCEEDED, cacheLocalPhoto),
  takeLatest(ActionTypes.NAVIGATE_TO_PHOTO_VIEWER, navigateToPhotoViewer),
]
