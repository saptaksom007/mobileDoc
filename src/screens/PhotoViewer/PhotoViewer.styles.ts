import { StyleSheet } from 'react-native'

export default StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#000'
  },
  transformImageActive: {
    flex: 1
  }
})
