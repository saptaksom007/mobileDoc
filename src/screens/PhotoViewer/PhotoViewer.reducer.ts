import {
  viewerRequested,
  viewerFailed,
  viewerSucceeded,
  viewerPop,
  viewerPreviewRequested,
  viewerPreviewSucceeded,
  initialState
} from 'utilities/commonViewerReducer'

import { ActionTypes } from './PhotoViewer.actions'
import { Action } from 'redux'

const {
  PHOTOVIEWER_REQUESTED,
  PHOTOVIEWER_SUCCEEDED,
  PHOTOVIEWER_FAILED,
  PHOTOVIEWER_PREVIEW_REQUESTED,
  PHOTOVIEWER_PREVIEW_SUCCEEDED,
  PHOTOVIEWER_POP
} = ActionTypes

export default function photoviewer(state: any = initialState, action: Action) {
  switch (action.type) {
    case PHOTOVIEWER_REQUESTED: {
      return viewerRequested(state, action)
    }

    case PHOTOVIEWER_SUCCEEDED: {
      return viewerSucceeded(state, action)
    }

    case PHOTOVIEWER_FAILED: {
      return viewerFailed(state)
    }

    case PHOTOVIEWER_PREVIEW_REQUESTED: {
      return viewerPreviewRequested(state, action)
    }

    case PHOTOVIEWER_PREVIEW_SUCCEEDED: {
      return viewerPreviewSucceeded(state, action)
    }

    case PHOTOVIEWER_POP: {
      return viewerPop(state)
    }

    default:
      return state
  }
}
