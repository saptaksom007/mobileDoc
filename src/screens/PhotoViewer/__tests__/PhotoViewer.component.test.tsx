import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import { PhotoViewer } from '../PhotoViewer.component'

jest.mock('ex-react-native-i18n', () => ({
  t: (key: string) => key
}))

jest.mock(
  '@applications-developer/react-native-transformable-image',
  () => 'TransformableImage'
)

it('<PhotoViewer/> render correctly', () => {
  const tree = renderer
    .create(<PhotoViewer dataUrl="" dispatch={() => {}} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
