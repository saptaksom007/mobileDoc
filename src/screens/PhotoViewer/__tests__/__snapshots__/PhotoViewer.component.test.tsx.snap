// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhotoViewer/> render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#000",
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "flex": 1,
          "justifyContent": "center",
        },
        Object {
          "backgroundColor": "transparent",
        },
      ]
    }
  >
    <ActivityIndicator
      animating={true}
      color="#fff"
      hidesWhenStopped={true}
      size="large"
    />
    <Text
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "fontFamily": "TitilliumWeb-Regular",
            "fontSize": 17.5,
            "lineHeight": 22,
            "overflow": "visible",
          },
          Object {
            "color": "rgba(27, 106, 158, 0.85)",
            "fontSize": 15,
            "lineHeight": 18,
          },
          undefined,
        ]
      }
    >
      
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "height": 166.75,
        "justifyContent": "center",
        "left": 0,
        "opacity": 1,
        "position": "absolute",
        "top": 0,
        "width": 62.5,
      }
    }
  />
</View>
`;
